File: 02-tcp/tcpServer.py
#!/usr/bin/env python

"""
    Python TCP echo server program for ANU COMP3310.

    Run with
        python tcpServer.py [ IP addr ] [ port ]

    Written by H Fisher u9011925, ANU, 2024
    This code may be freely copied and modified
"""

import sys
import socket
# Keep the code short for this tiny program
from socket import *

# Shared by client and server
from sockLine import readLine, writeLine


# IP address and port
serviceHost = "0.0.0.0"
servicePort = 3310


def serverLoop(host, port):
    """Run service on given host and port"""
    # Create a TCP socket
    serverSock = socket(AF_INET, SOCK_STREAM)
    # Address and port clients will connect to
    serverSock.bind((host, port))
    # This is the limit on ESTABLISHED connections that have not yet been
    # accepted. For now, it's magic you don't need to worry about.
    serverSock.listen(5)
    print("Created server socket for", serverSock.getsockname()[0],
                                        serverSock.getsockname()[1])
    # And now read and respond forever
    # while True:
    #     try:
    #         message, sender = readRequest(sock)
    #         replyToMessage(sock, message, sender)
    #     except OSError:
    #         break
    print("Close server socket")
    serverSock.close()


def processArgs(argv):
    """Handle command line arguments"""
    global serviceHost, servicePort
    #
    # This program has only two CLI arguments, and we know the order.
    # For any program with more than two args, use a loop or look up
    # the standard Python argparse library.
    if len(argv) > 1:
        serviceHost = argv[1]
        if len(argv) > 2:
            servicePort = int(argv[2])

##

if __name__ == "__main__":
    processArgs(sys.argv)
    serverLoop(serviceHost, servicePort)
    print("Done.")

File: 02-tcp/TcpServer.java

/** TCP echo server program for ANU COMP3310.
 *
 *  Run with
 *      java TcpServer [ IP addr ] [ port ]
 *
 *  Written by Hugh Fisher u9011925, ANU, 2024
 *  Released under Creative Commons CC0 Public Domain Dedication
 *  This code may be freely copied and modified for any purpose
 */


import java.io.*;
import java.net.*;


public class TcpServer {

    // IP address and port
    static String   serviceHost = "0.0.0.0";
    static int      servicePort = 3310;

    /** Accept client connections on given host and port */

    protected static void clientLoop(String host, int port)
        throws UnknownHostException, IOException, SocketException
    {
        ServerSocket        serverSock;
        Socket              client;
        InetSocketAddress   remote;

        // Create TCP socket, for server bound to given port and host.
        // Second arg is the limit on established TCP connections that have not yet
        // been accepted. (Not the number of connections.) A really busy server
        // might need to increase this but for now, don't worry about it.
        serverSock = new ServerSocket(port, 5, InetAddress.getByName(host));
        // Servers should set this option which allows them to be re-run immediately.
        // If you don't, you may have to wait a few minutes before restarting the server.
        serverSock.setReuseAddress(true);
        System.out.printf("Created server socket for %s %d\n",
                            serverSock.getInetAddress().getHostAddress(),
                            serverSock.getLocalPort());
        while (true) {
            // A TCP server is different to UDP. Instead of packets,
            // we get connection requests from clients.
            try {
                client = serverSock.accept();
            } catch (IOException e) {
                // If something goes wrong with the network, we will stop
                System.out.println(e.toString());
                break;
            }
            remote = (InetSocketAddress) client.getRemoteSocketAddress();
            System.out.printf("Accepted client connection from %s %d\n",
                            remote.getAddress().getHostAddress(), remote.getPort());
            // Each connection accepted creates a new socket for that
            // particular client. Use for requests and replies.
            //serverLoop(client);
            // We don't get back here until the client session ends
            client.close();
        }
        System.out.println("Close server socket");
        serverSock.close();
    }


    /** Handle command line arguments. */

    protected static void processArgs(String[] args)
    {
        //  This program has only two CLI arguments, and we know the order.
        //  For any program with more than two args, use a loop or package.
        if (args.length > 0) {
            serviceHost = args[0];
            if (args.length > 1) {
                servicePort = Integer.parseInt(args[1]);
            }
        }
    }

    public static void main(String[] args)
    {
        try {
            processArgs(args);
            clientLoop(serviceHost, servicePort);
            System.out.println("Done.");
        } catch (Exception e) {
            System.out.println(e.toString());
            System.exit(-1);   
        }
    }

}
